### ISSUE TITLE
Implement robust 7-segment digit recognition using a portable CNN (TensorFlow Lite)

### ISSUE THREAD

**Issue by rykhalskyi** [rykhalskyi/underpressure#22]
Background
The application must read numeric values from medical devices (e.g., blood pressure monitors) that display results on 7‑segment LCD screens.

The current approach uses Google ML Kit Text Recognition, which performs well for printed text but fails to reliably detect large LCD digits composed of separated segments.

Attempts to improve OCR accuracy using image preprocessing (thresholding, dilation, contrast enhancement) did not produce reliable results and were difficult to tune across varying lighting conditions.

Because the digits are not standard typographic glyphs, OCR‑based approaches are fundamentally limited.

A more reliable approach is to treat the problem as image classification of individual digits using a lightweight convolutional neural network.

Goal
Implement a robust digit recognition pipeline using a portable CNN model running locally on Android.

Requirements:

≥95% digit recognition accuracy

Fully offline operation

Processing time <200 ms

Robust to lighting and small camera angle variations

Minimal model size (<1 MB)

Proposed Solution
Use TensorFlow Lite to run a lightweight CNN digit classifier directly on the device.

The CNN will classify each digit independently.

Architecture
Код
Camera frame
   ↓
Detect LCD display
   ↓
Image preprocessing
   ↓
Digit segmentation
   ↓
CNN digit classifier (TensorFlow Lite)
   ↓
Combine digits into final values
Implementation Steps
1. Integrate TensorFlow Lite into the Android project
Add dependencies:

gradle
implementation 'org.tensorflow:tensorflow-lite:2.14.0'
implementation 'org.tensorflow:tensorflow-lite-support:0.4.4'
2. Add digit classifier model
Include a small CNN model file:

Код
digit_classifier.tflite
Place it in:

Код
app/src/main/assets/
Recommended characteristics:

Parameter	Value
Input size	28×28 pixels
Input type	grayscale
Classes	digits 0–9
Model size	50–200 KB
A standard MNIST‑style CNN is sufficient.

3. Detect LCD display region
Use either:

fixed crop region (if camera framing is controlled)

simple bounding‑box detection

Output:  
A Bitmap containing only the LCD screen.

4. Preprocess the image
Steps:

Convert to grayscale

Apply binary threshold

Remove noise using morphological filtering

The result should contain clear black digits on a white background.

5. Segment individual digits
Detect separate digit regions.

Possible techniques:

connected component detection

contour detection

bounding box filtering

Filtering criteria:

Parameter	Typical value
aspect ratio	0.3–0.8
min height	>30 px
max height	
Sort detected digit boxes left‑to‑right.

Example:

Код
[1] [5] [6]
6. Normalize digit images
Each detected digit should be:

cropped

centered

resized

Final format:

Код
28 × 28 grayscale image
pixel values 0.0–1.0
7. Run CNN inference
Load model:

java
Interpreter interpreter =
    new Interpreter(loadModelFile("digit_classifier.tflite"));
Prepare input tensor:

java
float[][][][] input = new float[1][28][28][1];
Run inference:

java
float[][] output = new float[1][10];
interpreter.run(input, output);
Predicted digit = index of highest probability.

8. Combine digits
Example:

Код
digits: [1,5,6]
Construct value:

Код
156
Repeat for:

SYS

DIA

PULSE

Performance Expectations
Typical mobile inference performance:

Metric	Expected Value
Model size	~100 KB
Inference time per digit	1–5 ms
Total processing time	
Testing Plan
Prepare a dataset of at least 50–100 device photos including:

different lighting conditions

slight camera angles

multiple blood pressure readings

different device models

Measure:

digit recognition accuracy

full reading accuracy

processing latency

Acceptance Criteria
The implementation is successful if:

digit classification accuracy ≥95%

full reading accuracy ≥90%

processing time ≤200 ms

works offline on Android devices

Risks
Potential issues:

poor digit segmentation under extreme lighting

reflections on LCD screens

non‑standard display layouts

Mitigation:

improve preprocessing

restrict detection to known display areas

validate numeric ranges

Deliverables
TensorFlow Lite model integration

digit segmentation module

digit classifier pipeline

performance benchmark results

documentation for future maintenance

Expected Outcome
A robust digit recognition system capable of reliably extracting values from 7‑segment LCD medical device displays, significantly outperforming OCR‑based approaches.
